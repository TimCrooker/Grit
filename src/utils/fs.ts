import fs, { promises as fsP } from 'fs'
import move from 'move-file'
import path from 'path'

export { ensureDir, remove, outputFile } from 'majo'

export const pathExists = async (path: string): Promise<boolean> => {
	try {
		await fs.promises.access(path)
		return true
	} catch (_) {
		return false
	}
}

export const copy = async (
	sourcePath: string,
	destinationPath: string,
	{ overwrite = true, directoryMode } = {}
): Promise<void> => {
	if (!sourcePath || !destinationPath) {
		throw new TypeError('`sourcePath` and `destinationPath` required')
	}

	if (!overwrite && (await pathExists(destinationPath))) {
		throw new Error(`The destination file exists: ${destinationPath}`)
	}

	await fsP.mkdir(path.dirname(destinationPath), {
		recursive: true,
		mode: directoryMode,
	})

	try {
		await fsP.rename(sourcePath, destinationPath)
	} catch (error) {
		if (error.code === 'EXDEV') {
			await fsP.copyFile(sourcePath, destinationPath)
			await fsP.unlink(sourcePath)
		} else {
			throw error
		}
	}
}

export { move }

export const readFile = fs.promises.readFile
