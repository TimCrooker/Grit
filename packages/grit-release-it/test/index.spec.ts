import path from 'path'
import { getGenerator, Grit } from 'grit-cli'
import fs from 'fs'

const generator = path.join(__dirname, '..')

test('defaults', async () => {
	const grit = await getGenerator({
		generator,
		mock: true,
		answers: { 'npm-publish': true, 'github-release': true, 'git-push': true }
	})

	fs.mkdirSync(grit.outDir, { recursive: true })
	fs.writeFileSync(path.join(grit.outDir, 'package.json'), JSON.stringify({}))

  await grit.run()
	
	expect(await grit.getOutputFiles()).toMatchSnapshot()

	const packageFile = await grit.readOutputFile('package.json') as Record<string, any>
  expect(packageFile.devDependencies["release-it"]).toEqual("^14.11.8")
	expect(packageFile.scripts.release).toEqual("release-it")
	
	const releaseFile = await grit.readOutputFile('.release-it.json') as Record<string, any>
	expect(releaseFile.github).toBeDefined()
	expect(releaseFile.git).toBeDefined()
})
