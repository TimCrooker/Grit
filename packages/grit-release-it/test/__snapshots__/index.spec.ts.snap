// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaults 1`] = `
Array [
  ".release-it.json",
  "package.json",
]
`;
